contracts/Market.sol

```diff
@@ -1929,6 +1771,168 @@ contract Market is
+        .getNextRemovalForSaleFromVintages({vintages: vintages});
+      // if removalId is 0 here, this supplier doesn't have any removals from the vintages requested
+      // at that point we should simply move on to the next supplier
+      if (removalId == 0) {
+        _incrementCurrentSupplierAddress();
+        continue;
+      }
+      uint256 removalAmount = _removal.balanceOf({
+        account: address(this),
+        id: removalId
+      });
+      if (remainingAmountToFill < removalAmount) {
+        /**
+         * The order is complete, not fully using up this removal, don't check about removing active supplier.
+         */
+        ids[countOfRemovalsAllocated] = removalId;
+        amounts[countOfRemovalsAllocated] = remainingAmountToFill;
+        suppliers[countOfRemovalsAllocated] = _currentSupplierAddress;
+        remainingAmountToFill = 0;
+      } else {
+        /**
+         * We will use up this removal while completing the order, move on to next one.
+         */
+        ids[countOfRemovalsAllocated] = removalId;
+        amounts[countOfRemovalsAllocated] = removalAmount; // this removal is getting used up
+        suppliers[countOfRemovalsAllocated] = _currentSupplierAddress;
+        remainingAmountToFill -= removalAmount;
+        /**
+        This may remove the supplier from the active suppliers and increment the current supplier address
+        * which is behavior we want in the case that the supplier is out of supply, but otherwise we do not
+        * interact with the global _currentSupplierAddress in this fulfillment flow.
+        */
+        _removeActiveRemoval({
+          removalId: removalId,
+          supplierAddress: _currentSupplierAddress
+        });
+      }
+      ++countOfRemovalsAllocated;
+      if (remainingAmountToFill == 0) {
+        break;
+      }
+    }
+    return (countOfRemovalsAllocated, ids, amounts, suppliers);
+  }
+
+  /**
+   * @notice Allocates the removals, amounts, and suppliers needed to fulfill the purchase drawing only from
+   * the specific supplier and vintages specified.
+   * @param amount The number of carbon removals to purchase.
+   * @param supplier The supplier from which to purchase carbon removals.
+   * @param vintages The vintages from which to purchase carbon removals.
+   * @return countOfRemovalsAllocated The number of distinct removal IDs used to fulfill this order.
+   * @return ids An array of the removal IDs being drawn from to fulfill this order.
+   * @return amounts An array of amounts being allocated from each corresponding removal token.
+   */
+  function _allocateSupplySingleSupplierSpecificVintages(
+    uint256 amount,
+    address supplier,
+    uint256[] memory vintages
+  )
+    private
+    returns (
+      uint256 countOfRemovalsAllocated,
+      uint256[] memory ids,
+      uint256[] memory amounts
+    )
+  {
+    RemovalsByYear storage supplierRemovalQueue = _listedSupply[supplier];
```

Comment:
Remind me why this needs to be storage?

Reply:
because this is a storage variable -- we need any mutations to this variable to persist in state